### DISCUSSION TITLE
How do I get the filename if I'm using bentoml.io.File API descriptor for the input?

### DISCUSSION THREAD

**Discussion by marygmalana** [bentoml/BentoML#2381] [Q&A] [answered]
I understand that the input/output becomes a Filelike object. I want to get the filename because I will use it for post processing.

I tried calling input.name but it returns None. I also tried input.path but I get an "TypeError: Expected str, got bytes" on "raw_path = urllib.request.url2pathname(urllib.parse.unquote(parsed.path))" line. 

It seems very obvious but I cant seem to find the way to get the filename. Thank you in advance for your guidance!

**Comment by sauyon** [comment, marked as answer]
Hi! There currently isn't a way to get a filename, as there is no actual file on disk. If you need to pass a filename to something unfortunately you'll have to write it manually first.

**Comment by marygmalana** [comment]
@sauyon thank you for reply! I tried to use the Multipart API descriptor so that I can write the file name and also upload the data. I used the following code 
```
input_spec = Multipart(vid=File(), filename=Text())
@svc.api(input=input_spec, output=JSON())
```
However, I got a BentoMLException: _File should have Content-Type b'application/octet-stream' or  b'multipart/form-data', got b'video/mp4' instead_. I tried using multipart with other descriptor like NumpyNdarray, it did not raise an error. What do you think?

**Reply by aarnphm** [reply]
for parsing a video, currently, we only accept the `Content-Type` in header that must be either `octet-stream` or `form-data` for `File` IO descriptor. https://github.com/bentoml/BentoML/blob/c6a50e6bfc5f4f2be7ce2e4bd793fcd776a67e9e/bentoml/_internal/io_descriptors/file.py#L98 https://github.com/bentoml/BentoML/blob/c6a50e6bfc5f4f2be7ce2e4bd793fcd776a67e9e/bentoml/_internal/io_descriptors/file.py#L93

**Reply by marygmalana** [reply]
@aarnphm,  does that mean I have to change the request format? When I use _@svc.api(input=File(), output=JSON())_, I don't encounter any error with the video. When I changed it to MultiPart, I get the error that I mentioned above. 

Thank you very much for your help! Currently, I just use the Swagger UI to test out the API endpoints.

**Reply by parano** [reply]
@aarnphm perhaps File should allow other content-type as well?

**Reply by aarnphm** [reply]
I think we should introduce a `Video` IO type instead. I prefer `File` to be accepted as a stream. But the argument is what is really the benefits of introducing a `Video` IO type? Unless we want to do sth special I think for most cases file can converted to byte stream @parano?

**Reply by aarnphm** [reply]
> @aarnphm, does that mean I have to change the request format? When I use _@svc.api(input=File(), output=JSON())_, I don't encounter any error with the video. When I changed it to MultiPart, I get the error that I mentioned above.

This means in your curl command you can do sth like
```bash
curl -X POST -H "Content-Type: multipart/form-data" -F annotations=@test.json -F arr='[5,4,3,2]'
```

**Reply by marygmalana** [reply]
@aarnphm I'm sorry I am not able to follow. First, I need to convert the video to byte stream so that I can use it with multipart. How do I convert the video to byte stream?

**Reply by aarnphm** [reply]
sorry for the confusion, what I meant is you can just set the header to form-data. You can also stream your video as binary this way
```bash
curl --data-binary ...
```
@mgmalana

**Reply by marygmalana** [reply]
@aarnphm thank you for your explanation. I was able to fix it by removing "type=video/mp4" from the curl command. I was using "vid=@test.mp4;type=video/mp4". Now it works as expected!

**Reply by sauyon** [reply]
> I think we should introduce a `Video` IO type instead. I prefer `File` to be accepted as a stream. But the argument is what is really the benefits of introducing a `Video` IO type? Unless we want to do sth special I think for most cases file can converted to byte stream @parano?

I think it would be valuable, because users that wish to only accept video content might have something use the `Video` descriptor to
 a) describe that what they're looking for is an API endpoint that takes video, maybe giving us hints at serving-time what kinds of inputs we might expect,
 b) document to other developers that the API endpoint takes video, and
 c) restrict the API endpoint to content-types related to video, maybe with options like `Video(kind='mp4')` to limit to specific video formats.

IMO the `File` descriptor should probably just be a catch-all which accepts any content type, and then we'd have subclasses like `Video` which specifically restrict the content type.

I've opened https://github.com/bentoml/BentoML/issues/2387 in case we want to address situations like this in the future.
